### ISSUE TITLE
Intergrated to Transformer Engine

### ISSUE THREAD

**Issue by shuxiaobo** [OpenNLPLab/lightning-attention#21]
hi, Is There Any Lightning Attention Intergrated to Transformer Engine Right Now?

**Comment by Doraemonzzz** [comment]
Are you saying that there is code available that can be directly used for training? Or a model that is compatible with the HF interface?

**Comment by shuxiaobo** [comment]
Yes, for training code of LLM Infra，like Megatron or deepspeed (transformer engine embeded)

**Comment by Doraemonzzz** [comment]
Thank you for your suggestion. I will provide a version similar to nanogpt in the coming days to facilitate developers' experimentation. Once there are updates, I will update here.

**Comment by shuxiaobo** [comment]
thanks very much! is there wechat for discuss some insight of lightning? This is my wechat-id: wxid_c8qfx6yqdw522

**Comment by shuxiaobo** [comment]
I still can not understand Attention Mask in lightning, `AttentionMask & PositonEmbedding`, what does `_slope` means in the code(it seems \lambda in paper)? How does attn mask be implement in triton code?

**Comment by Doraemonzzz** [comment]
$\lambda = \exp(slope)$, you can take [this line](https://github.com/OpenNLPLab/lightning-attention/blob/main/lightning_attn/ops/triton/lightning_attn2.py#L54) for reference.

**Comment by shuxiaobo** [comment]
@Doraemonzzz The last norm of lightning is on the `hidden_size` dim? does it work on the `num_hidden_per_head`？

**Comment by Doraemonzzz** [comment]
This actually corresponds to two versions. In one version, we first perform rearrange and then normalize; in the other, we normalize first and then rearrange. The former is used in TNL, and it achieves better performance than the latter (as confirmed by my recent experiments).

**Comment by shuxiaobo** [comment]
I have done some dense model exp using lightning, which with 2/24, 12/24 layer full-attn. loss is better than pure dense llama🎉, but the speed is slow than Dense model with the same number of parameters, which is 6.7s(pure dense) vs 8.5s(2la per 24layer) per iteration. 
megatron backend, no parallel mechanism, on flash v2 attn, 64GPUs, 24layer with hidden 3072 ffn 8192, is that normal?
